src/algebra/char_p/mixed_char_zero.lean

```diff
@@ -0,0 +1,452 @@
+  [fact (∀ (I : ideal R), I ≠ ⊤ → char_zero (R ⧸ I))] (n : ℕ+) :
+  ((n : Rˣ) : R) = ↑n :=
+begin
+  change ((equal_char_zero.pnat_coe_is_unit R n).unit : R) = ↑n,
+  simp only [is_unit.unit_spec],
+end
+
+/--
+Equal characteristic implies `ℚ`-algebra.
+-/
+noncomputable def equal_char_zero_to_Q_algebra (h : ∀ (I : ideal R), I ≠ ⊤ → char_zero (R ⧸ I)) :
+  algebra ℚ R :=
+begin
+  -- turning `h` into a `fact` since we used it that way in the helper lemmas above.
+  haveI : fact (∀ (I : ideal R), I ≠ ⊤ → char_zero (R ⧸ I)) := by {rw fact_iff, exact h},
+
+  apply ring_hom.to_algebra,
+  exact
+  { to_fun := λ x, x.num /ₚ ↑(x.pnat_denom),
+    map_zero' := by simp [divp],
+    map_one' := by simp [equal_char_zero.pnat_coe_units_eq_one],
+    map_mul' :=
+    begin
+      intros a b,
+      field_simp,
+      repeat { rw equal_char_zero.pnat_coe_units_coe_eq_coe R },
+      convert_to (↑((a * b).num * (a.denom) * (b.denom)) : R) = _,
+      { simp_rw [int.cast_mul, int.cast_coe_nat, coe_coe, rat.coe_pnat_denom],
+        ring },
+      rw rat.mul_num_denom' a b,
+      simp
+    end,
+    map_add' :=
+    begin
+      intros a b,
+      field_simp,
+      repeat { rw equal_char_zero.pnat_coe_units_coe_eq_coe R },
+      convert_to (↑((a + b).num * a.denom * b.denom) : R)  = _,
+      { simp_rw [int.cast_mul, int.cast_coe_nat, coe_coe, rat.coe_pnat_denom],
+        ring },
+      rw rat.add_num_denom' a b,
+      simp
+    end }
+end
+
+end construction_of_Q_algebra
+
+end equal_char_zero
+
+/--
+Not mixed characteristic implies equal characteristic.
+-/
+lemma not_mixed_char_to_equal_char_zero [char_zero R] (h : ¬(∃ p > 0, mixed_char_zero R p)) :
+  ∀ (I : ideal R), I ≠ ⊤ → char_zero (R ⧸ I) :=
+begin
+  push_neg at h,
+  intros I hI_ne_top,
+  apply char_p.char_p_to_char_zero _,
+  cases char_p.exists (R ⧸ I) with p hp,
+  cases p,
+  { exact hp },
+  { have h_mixed : mixed_char_zero R p.succ := ⟨⟨I, ⟨hI_ne_top, hp⟩⟩⟩,
+    exact absurd h_mixed (h p.succ p.succ_pos) }
+end
+
+example (n : ℕ) (h : 0<n) : n ≠ 0 := ne_of_gt h
+
+/--
+Equal characteristic implies not mixed characteristic.
+-/
+lemma equal_char_zero_to_not_mixed_char (h : ∀ (I : ideal R), I ≠ ⊤ → char_zero (R ⧸ I)) :
+  ¬(∃ p > 0, mixed_char_zero R p) :=
+begin
+  push_neg,
+  intro p,
+  by_contradiction hp,
+  push_neg at hp,
+  let hp_pos := hp.1,
+  let hp_mixed_char := hp.2,
+  rcases hp_mixed_char.char_p_quotient with ⟨I, ⟨hI_ne_top, hI_p⟩⟩,
+  haveI hI_zero : char_zero (R ⧸ I) := (h I hI_ne_top),
+  replace hI_zero : char_p (R ⧸ I) 0 := char_p.of_char_zero _,
+  exact absurd (char_p.eq (R ⧸ I) hI_p hI_zero) (ne_of_gt hp_pos),
+end
+
+/--
+A ring of characteristic zero has equal characteristic iff it does not
+have mixed characteristic for any `p`.
+-/
+lemma equal_char_zero_iff_not_mixed_char [char_zero R] :
+  (∀ (I : ideal R), I ≠ ⊤ → char_zero (R ⧸ I)) ↔ (¬(∃ p > 0, mixed_char_zero R p)) :=
+iff.intro (equal_char_zero_to_not_mixed_char R) (not_mixed_char_to_equal_char_zero R)
```

Comment:
We typically use these pointy brackets instead of naming the constructors like `iff.intro`:
```suggestion
⟨equal_char_zero_to_not_mixed_char R, not_mixed_char_to_equal_char_zero R⟩
```